### PR TITLE
Implementation of Attributes-per-Tag via Interfaces

### PR DESCRIPTION
**Pull request by pointbazaar**
…efault methods. Drawback: 'get()' method that does nothing, typecasting and many classes and interfaces

### CODE REVIEW

**Comment by pointbazaar** [comment]
Can i get some feedback on this implementation?
If you like it, i could try to work down the list of attributes [Attributes](https://www.w3schools.com/tags/ref_attributes.asp)
in this draft PR until all the attributes are implemented. 

What bothers me the most is not being able to get a type of the actual class instance when implementing default methods in an 
interface. Maybe there is another way? 
The way I'm doing it now to is to use the type parameter of the interface to cast 'this' to 'T',
so the default method in the interface can return the correct type.

Currently the type of the tags is also watered down to the type they are extending when using any attributes that are not specific to the tag. This could be avoided by maybe generifying the ContainerTag. But i wanted to get feedback on this direction first, as there would be quite the number of new classes and interfaces.

**Comment by pointbazaar** [comment]
Now the TagCreatorCodeGenerator will generate methods that return the custom type for each tag.

**Comment by pointbazaar** [comment]
- EmptyTag generified
- ContainerTag generified
- generate interfaces differently depending on if the attribute has an argument or not

What remains to do:
- double-check the attributes and which tags they apply to
- generate the interfaces in a way that is consistent with the rest of the project (like isHidden(), withCondHidden(...) ), because currently it is just one method e.g. withDownload(final boolean ...) which is more to write

**Comment by pointbazaar** [comment]
- added all **global attributes** in Tag.java
- changed the Code Generator for the Interfaces to generate methods consistent with the rest of the project.
- pruned attribute methods from Tag.java that were not valid on all HTML tags

What remains:
- checking ContainerTag.java and EmptyTag.java for extra methods
- using correct default attribute values for attributes without argument e.g. ```.isAutoFocus()``` should generate ```<$TAG autofocus>```... 
but ```.isAutocomplete()``` should generate ```<$TAG autocomplete="on">``` ...

This can be adjusted in AttributeList.java vie AttrD constructor.

**Comment by pointbazaar** [comment]
- Wrote Tests for Attributes-Per-Tag, covering some of the generated methods
- Went through the Attributes list from W3Schools and enter the attribute property if they have an argument or not

Probably some special cases have been missed and also some attributes have not even been included here, most of them because of a dash ('-') in their name, which would force me to make some decisions as to how their interfaces and attribute-specific methods should be named.

I'm facing some **tradeoffs** with this PR currently:
- writing many tests for attributes-per-tag would mean wasted effort if there is a decision
 to change the syntax later and would make it complicated (hand-refactoring and then re-generating the interfaces would be required if one would not want to rewrite those tests)
- it would be **very convenient** to have special ```with$ATTR(int $ATTR) ``` Methods for example for ```withTabindex``` where one could enter a number instead of a string. Because the attribute is a number already. This would spare people generating their HTML the ```x+""```, all too common in some codebases. However this would complicate code generation even further, and it already looks like spaghetti code and could use some refactoring.

However i am confident that this PR in it's current state represents an improvement in type-safety and convenience for users of J2HTML and would like to have it reviewed. It has been going on for too long and has collected too many commits already.
The gaps left by this PR can easily be added in later on and in my opinion feel like an omission and not a regression for the API.

**Comment by tipsy** [comment]
It's gotten quite big, yeah. I will go through it in the weekend.

File: .gitignore [hunk 1]
```diff
@@ -1,5 +1,6 @@
 # Created by https://www.gitignore.io
 
+src/main/java/j2html/Experiment.java
```
Comment (tipsy) [inline .gitignore:3]: This seems like something that shouldn't be tracked.

**Comment by tipsy** [comment]
I didn't fully read through the generator code, but the generated code looks good.

It would be helpful to add a section in the readme, or create a `CONTRIBUTING.md` file to describe the new architecture and the role of the generators, as everything is a bit more complex now. This can be done in a separate PR though.

File: .gitignore (hunk 1 above)
Comment (pointbazaar) [reply .gitignore]: yeah, i forgot i used that one for trying out stuff.

**Comment by pointbazaar** [comment]
[PR](https://github.com/tipsy/j2html/pull/157) is out!

**Comment by tipsy** [comment]
Great work :)

**Pull request merged**